metecho/api/views.py

```diff
@@ -58,6 +58,8 @@ def perform_create(self, serializer):
         return super().perform_create(serializer)
 
     def perform_update(self, serializer):
+        # TODO: readonly users can assign/unassign themselves as `task.assigned_qa`
+        # TODO: readonly users can remove themselves from `epic.github_users` list
```

Comment:
@jerivas These are the two pieces where I'd like your help. Removing `enforce_push_permission=True` from the serializer for `assigned_qa` got me part of the way, but we also want readonly users to be able to make limited updates to tasks and epics:

- task: add/remove themselves (and only themselves) as `assigned_qa`
- epic: remove themselves from `github_users` (I suppose allowing adding is okay too, though I don't think we'll expose that via the UI)

Reply:
What do you think of moving assignee/collaborator logic to a separate endpoint? Something like:

```http
POST or DELETE /api/tasks/<id>/assignees/

{
  "role": "qa",
  "user": "abc123",
}
```

And for Epics:

```http
POST or DELETE /api/epics/<id>/collaborators/

{
   "user": "abc123"
}
```

I propose this because I sense the permission checks will get pretty wild if we allow/deny updates based on the fields the user is trying to modify. Fully authorized users can use this new endpoint as well if we don't want the frontend to hit different URLs depending on permissions.

Reply:
Actually, we might be able to get away with it if we `PATCH` instead of `POST` on the regular endpoints, that should make it easier to detect if a user is updating **just** the `assigned_*` or `github_users` fields. That might be best if separate endpoints are too annoying to implement in the frontend.

Reply:
@jerivas I'm honestly fine with either approach. Both are relatively straightforward in the JS, so I'd defer to whichever you think will be cleaner for the API. (I'd lean toward all users using the same endpoint, though, and allowing the API to determine permissions.)

Reply:
Alright, turns out handling the assignees separately streamlined the vanilla `TaskSerializer` and required very minimal frontend changes. Everything works in my manual tests, just need to backfill tests and do something similar for collaborators

Reply:
As of https://github.com/oddbird/Metecho/pull/1106/commits/85fb547b414aa14ada1c75c93516d0bd9e045bc4:

- Added `epic-collaborators` endpoint
- Added tests for both new endpoints